worker/upgradedatabase/worker.go

```diff
@@ -280,9 +283,31 @@ func (w *upgradeDB) watchUpgrade() {
 	if err := w.upgradeInfo.Refresh(); err != nil {
 		w.logger.Errorf("unable to refresh upgrade info: %v", err)
 		w.setFailStatus()
-		return
+		return err
 	}
 
+	// To be here, this node previously returned false for isPrimary
+	// Sometimes our primary changes, or is reported incorrectly as
+	// false. In the case that a node becomes primary whilst watching,
+	// esclate an error which will result in the worker being restarted
+	becamePrimary := make(chan struct{}, 1)
+	done := make(chan struct{}, 1)
+	defer close(done)
+	go func() {
+		defer close(becamePrimary)
+		for {
+			select {
+			case <-done:
+			case <-w.clock.After(5 * time.Second):
+				isPrimary, err := w.pool.IsPrimary(w.tag.Id())
+				if err == nil && !isPrimary {
```

Comment:
The way I have resolved this is by changing to `err != nil || !isPrimary` so you only exit, marking this host as becamePrimary is IsPrimary() returns true with no error.

Only problem with this approach is errors returned from IsPrimary() in this goroutine are ignored. I suspect this isn't a problem though, since we don't really care

Reply:
The way I have resolved this is by changing to `err != nil || !isPrimary` so you only exit, marking this host as becamePrimary is IsPrimary() returns true with no error.

Only problem with this approach is errors returned from IsPrimary() in this goroutine are ignored. I suspect this isn't a problem though, since we don't really care

Reply:
It turns out that when the primary state changes, this can (and usually does) result in IsPrimary throwing an error (and returning false) so my original fix to this will not work

So instead, I've reverted to the original, and changed wording to signal a 'state change' instead of becoming primary (which is uncertain)

Reply:
I'd make this more terse with:
```go
if isPrimary, _ := w.pool.IsPrimary(w.tag.Id()); isPrimary {
    close(stateChanged)
    return
}
```
and drop the deferred close above, and the continue + return below.

Reply:
OK, I'd still go with something like:
```go
isPrimary, err := w.pool.IsPrimary(w.tag.Id())
if err != nil {
    w.logger.Errorf(err)
    close(stateChanged) 
    return 
}
if isPrimary { 
    close(stateChanged) 
    return 
}
```
And probably just delete the log line on 330.